Cabal/src/Distribution/Simple/Haddock.hs

```diff
@@ -563,6 +573,11 @@ runHaddock verbosity tmpFileOpts comp platform haddockProg args
 
         notice verbosity $ "Documentation created: " ++ result
 
+        when (fromFlag (argOpen args)) $ do
+          exe <- findExecutable "open" >>= maybe (findExecutable "xdg-open") (pure . pure)
```

Comment:
* I think this should depend on the platform.
* I'm not sure whether this should use the whole `Distribution/Simple.Program` thing, but it's something to consider. It looks like no other place uses `findExecutable` directly, relying on `ProgramDB instead`. On the other hand, this is a `cabal-install` exclusive command, and I don't know if there's a way to keep the `ProgramDB` changes out of Cabal.

Reply:
Re. platform: I agree. It looks like I can check the `LocalBuildInfo` for the current platform, so that should work. I’ll dispatch to `xdg-open` on Linux and the BSDs. @gbaz pointed out that I didn’t account for a Windows case; [these docs](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/start) say that `start` should work.

Reply:
Good point. I'd still move all the open-file logic in a separate module, since it's independent from haddock stuff

Reply:
Moved the code that cases over the `OS` to `Distribution.System`, where it can live alongside the `Platform` type.